File: app/src/main/java/eu/darken/bluemusic/bluetooth/core/EventGenerator.java
package eu.darken.bluemusic.bluetooth.core;

import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;

import javax.inject.Inject;

import eu.darken.bluemusic.AppComponent;
import eu.darken.bluemusic.main.core.service.ServiceHelper;
import timber.log.Timber;

@AppComponent.Scope
public class EventGenerator {
    public static final String EXTRA_DEVICE_EVENT = "eu.darken.bluemusic.core.bluetooth.event";
    private final Context context;

    @Inject
    public EventGenerator(Context context) {
        this.context = context;
    }

    public SourceDevice.Event generate(SourceDevice device, SourceDevice.Event.Type type) {
        return new SourceDevice.Event(device, type);
    }

    public void send(SourceDevice device, SourceDevice.Event.Type type) {
        Timber.d("Generating and sending %s for %s", type, device);
        SourceDevice.Event event = generate(device, type);

        Intent service = ServiceHelper.getIntent(context);
        service.putExtra(EXTRA_DEVICE_EVENT, event);
        final ComponentName componentName = ServiceHelper.startService(context, service);
        if (componentName != null) Timber.v("Service is already running.");
    }
}

File: app/src/main/java/eu/darken/bluemusic/common/EventGenerator.kt
package eu.darken.bluemusic.common

import android.content.Context
import dagger.hilt.android.qualifiers.ApplicationContext
import eu.darken.bluemusic.bluetooth.core.SourceDevice
import eu.darken.bluemusic.common.debug.logging.log
import eu.darken.bluemusic.common.debug.logging.logTag
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class EventGenerator @Inject constructor(
    @ApplicationContext private val context: Context
) {

    companion object {
        private val TAG = logTag("EventGenerator")
        const val EXTRA_DEVICE_EVENT: String = "eu.darken.bluemusic.core.bluetooth.event"
    }

    fun generate(device: SourceDevice, type: SourceDevice.Event.Type): SourceDevice.Event {
        return SourceDevice.Event(device, type)
    }

    fun send(device: SourceDevice, type: SourceDevice.Event.Type) {
        log(TAG) { "Generating and sending $type for $device" }
        generate(device, type)
        // TODO
//        val service: Intent = ServiceController.getIntent(context)
//        service.putExtra(EXTRA_DEVICE_EVENT, event)
//        val componentName: ComponentName? = ServiceController.startService(context, service)
//        if (componentName != null) log(TAG, VERBOSE) { "Service is already running." }
    }
}
